File: apps/collector/src/alerts.ts
import type { VenueHealth } from "./health";

export interface AlertPayload {
  text: string;
  ok: boolean;
  stale: string[];
}

export type AlertSink = (payload: AlertPayload) => Promise<void>;

/** The only part of fetch the sink uses, so a test can pass a plain function. */
export type FetchLike = (url: string, init?: RequestInit) => Promise<Response>;

/**
 * Reports venues that stop collecting, and their recovery, over a webhook.
 *
 * Only transitions are sent: a message when venues newly go stale, and one when everything is
 * healthy again. A venue recovering while others are still stale stays quiet, so a partial outage
 * produces a couple of messages rather than one a minute until someone looks.
 */
export class StaleVenueAlerter {
  private stale: string[] = [];

  constructor(
    private readonly send: AlertSink,
    private readonly log?: (message: string) => void,
  ) {}

  async check(snapshot: { venues: readonly VenueHealth[] }): Promise<void> {
    const stale = snapshot.venues
      .filter((venue) => venue.stale)
      .map((venue) => venue.venueId)
      .sort();
    const newlyStale = stale.filter((venueId) => !this.stale.includes(venueId));
    const recovered = this.stale.length > 0 && stale.length === 0;
    const previous = this.stale;
    this.stale = stale;

    if (newlyStale.length > 0) {
      const errors = snapshot.venues
        .filter((venue) => venue.stale && venue.error)
        .map((venue) => `${venue.venueId}: ${venue.error}`);
      await this.report({
        ok: false,
        stale,
        text: [
          `airates: ${stale.length} venue${stale.length === 1 ? "" : "s"} stale (${stale.join(", ")})`,
          errors.length > 0 ? `last errors -- ${errors.join("; ")}` : null,
        ]
          .filter(Boolean)
          .join("\n"),
      });
      return;
    }

    if (recovered) {
      await this.report({
        ok: true,
        stale: [],
        text: `airates: all venues collecting again (was ${previous.join(", ")})`,
      });
    }
  }

  private async report(payload: AlertPayload): Promise<void> {
    this.log?.(payload.text);
    await this.send(payload);
  }
}

/** Posts to a Slack- or Discord-style incoming webhook; both field names are sent. */
export function webhookSink(url: string, fetchImpl: FetchLike = fetch): AlertSink {
  return async (payload) => {
    const response = await fetchImpl(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ ...payload, content: payload.text }),
      signal: AbortSignal.timeout(10_000),
    });
    if (!response.ok) throw new Error(`webhook returned HTTP ${response.status}`);
  };
}

File: apps/collector-go/internal/collector/alerts.go
package collector

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"time"
)

// AlertPayload is one message about collection health.
type AlertPayload struct {
	Text  string   `json:"text"`
	OK    bool     `json:"ok"`
	Stale []string `json:"stale"`
}

// AlertSink delivers a payload somewhere a person will see it.
type AlertSink func(ctx context.Context, payload AlertPayload) error

// StaleVenueAlerter reports venues that stop collecting, and their recovery, over a sink.
//
// Ported from apps/collector/src/alerts.ts. Only transitions are sent: a message when venues newly go
// stale, and one when everything is healthy again. A venue recovering while others are still stale
// stays quiet, so a partial outage produces a couple of messages rather than one a minute until
// someone looks.
//
// Not safe for concurrent use; it runs inside one PeriodicTask.
type StaleVenueAlerter struct {
	send  AlertSink
	log   func(string)
	stale []string
}

func NewStaleVenueAlerter(send AlertSink, log func(string)) *StaleVenueAlerter {
	return &StaleVenueAlerter{send: send, log: log}
}

// Check compares the fleet with the last check and reports any transition.
func (a *StaleVenueAlerter) Check(ctx context.Context, snapshot Snapshot) error {
	stale := []string{}
	var errors []string
	for _, venue := range snapshot.Venues {
		if !venue.Stale {
			continue
		}
		stale = append(stale, venue.VenueID)
	}
	sort.Strings(stale)
	for _, venue := range snapshot.Venues {
		if venue.Stale && venue.Error != nil && *venue.Error != "" {
			errors = append(errors, fmt.Sprintf("%s: %s", venue.VenueID, *venue.Error))
		}
	}

	previous := a.stale
	newlyStale := false
	for _, id := range stale {
		if !contains(previous, id) {
			newlyStale = true
			break
		}
	}
	recovered := len(previous) > 0 && len(stale) == 0
	a.stale = stale

	if newlyStale {
		plural := "s"
		if len(stale) == 1 {
			plural = ""
		}
		text := fmt.Sprintf("airates: %d venue%s stale (%s)", len(stale), plural, strings.Join(stale, ", "))
		if len(errors) > 0 {
			text += "\nlast errors -- " + strings.Join(errors, "; ")
		}
		return a.report(ctx, AlertPayload{Text: text, OK: false, Stale: stale})
	}
	if recovered {
		return a.report(ctx, AlertPayload{
			Text:  fmt.Sprintf("airates: all venues collecting again (was %s)", strings.Join(previous, ", ")),
			OK:    true,
			Stale: []string{},
		})
	}
	return nil
}

func (a *StaleVenueAlerter) report(ctx context.Context, payload AlertPayload) error {
	if a.log != nil {
		a.log(payload.Text)
	}
	return a.send(ctx, payload)
}

func contains(list []string, value string) bool {
	for _, item := range list {
		if item == value {
			return true
		}
	}
	return false
}

// WebhookSink posts to a Slack- or Discord-style incoming webhook. Both field names carry the text:
// Slack reads `text`, Discord reads `content`.
func WebhookSink(url string, client *http.Client) AlertSink {
	if client == nil {
		client = &http.Client{Timeout: 10 * time.Second}
	}
	return func(ctx context.Context, payload AlertPayload) error {
		body, err := json.Marshal(struct {
			AlertPayload
			Content string `json:"content"`
		}{payload, payload.Text})
		if err != nil {
			return fmt.Errorf("encode alert: %w", err)
		}
		request, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
		if err != nil {
			return fmt.Errorf("build alert request: %w", err)
		}
		request.Header.Set("content-type", "application/json")
		response, err := client.Do(request)
		if err != nil {
			return fmt.Errorf("post alert: %w", err)
		}
		defer response.Body.Close()
		if response.StatusCode < 200 || response.StatusCode > 299 {
			return fmt.Errorf("webhook returned HTTP %d", response.StatusCode)
		}
		return nil
	}
}
